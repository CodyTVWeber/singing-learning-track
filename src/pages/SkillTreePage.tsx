import React, { useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { useApp } from '../context/AppContext';
import { getAllUnits, isLessonUnlocked } from '../data/units';
import { Card } from '../components/Card';
import { Container } from '../components/Container';
import { colors, fontSize, fontWeight, spacing, borderRadius, shadows } from '../theme/theme';
import type { Lesson } from '../models/lesson';
import { analytics } from '../services/analytics';
import { Icon } from '../components/Icon';
import { Chip } from '../components/Chip';

export const SkillTreePage: React.FC = () => {
  const navigate = useNavigate();
  const { user, getCompletedLessonIds } = useApp();
  const units = getAllUnits();
  const completedLessonIds = getCompletedLessonIds();

  useEffect(() => {
    if (!user) {
      navigate('/onboarding', { replace: true });
    }
  }, [user, navigate]);

  const getLessonStatus = (lesson: Lesson) => {
    if (completedLessonIds.includes(lesson.id)) {
      return 'completed';
    }
    if (isLessonUnlocked(lesson.id, completedLessonIds)) {
      return 'unlocked';
    }
    return 'locked';
  };

  const handleLessonClick = (lesson: Lesson) => {
    const status = getLessonStatus(lesson);
    if (status !== 'locked') {
      navigate(`/lesson/${lesson.id}`);
    }
  };

  useEffect(() => {
    if (!user) return;
    analytics.trackEvent('map_loaded', {
      unitsCount: units.length,
      completedLessons: completedLessonIds.length,
    });
  }, [user, units.length, completedLessonIds.length]);

  if (!user) {
    return null;
  }

  return (
    <div
      style={{
        minHeight: '100vh',
        backgroundColor: colors.background,
        paddingBottom: spacing.xl,
      }}
    >
      {/* Header */}
      <div
        style={{
          backgroundColor: colors.primary,
          color: 'white',
          padding: spacing.lg,
          boxShadow: shadows.md,
        }}
      >
        <Container>
          <div
            style={{
              display: 'flex',
              justifyContent: 'space-between',
              alignItems: 'center',
              flexWrap: 'wrap',
              gap: spacing.md,
            }}
          >
            <div>
              <h1
                style={{
                  fontSize: fontSize.xl,
                  fontWeight: fontWeight.bold,
                  marginBottom: spacing.xs,
                  color: 'white',
                }}
              >
                Hi {user.name}!
              </h1>
              <p style={{ fontSize: fontSize.md, opacity: 0.9 }}>
                Keep singing every day!
              </p>
            </div>
            <div
              style={{
                display: 'flex',
                gap: spacing.lg,
                alignItems: 'center',
              }}
            >
              {/* Streak chip */}
              {typeof user.streakCount === 'number' && user.streakCount > 0 && (
                <Chip
                  label={`${user.streakCount} day${user.streakCount === 1 ? '' : 's'} ðŸ”¥`}
                  color="warning"
                  variant="filled"
                  size="medium"
                />
              )}
              <div style={{ textAlign: 'center' }}>
                <div style={{ fontSize: fontSize.xs, opacity: 0.8 }}>Points</div>
                <div style={{ fontSize: fontSize.xl, fontWeight: fontWeight.bold }}>
                  {user.totalPoints}
                </div>
              </div>
            </div>
          </div>
        </Container>
      </div>

      {/* Units */}
      <Container>
        {units.map((unit) => (
          <div key={unit.unit} style={{ marginTop: spacing.xl }}>
            <h2
              style={{
                fontSize: fontSize.xxl,
                fontWeight: fontWeight.bold,
                color: colors.darkBrown,
                marginBottom: spacing.md,
              }}
            >
              Unit {unit.unit}: {unit.title}
            </h2>
            <p
              style={{
                fontSize: fontSize.lg,
                color: colors.textLight,
                marginBottom: spacing.lg,
              }}
            >
              {unit.description}
            </p>

            <div
              style={{
                display: 'grid',
                gridTemplateColumns: 'repeat(auto-fill, minmax(280px, 1fr))',
                gap: spacing.lg,
              }}
            >
              {unit.lessons.map((lesson) => {
                const status = getLessonStatus(lesson);
                return (
                  <Card
                    key={lesson.id}
                    onClick={() => handleLessonClick(lesson)}
                    style={{
                      opacity: status === 'locked' ? 0.6 : 1,
                      cursor: status === 'locked' ? 'not-allowed' : 'pointer',
                      position: 'relative',
                      overflow: 'hidden',
                    }}
                  >
                    {/* Transparent overlay to track clicks */}
                    <div
                      onClick={() => {
                        if (status !== 'locked') {
                          analytics.trackEvent('lesson_card_click', {
                            lessonId: lesson.id,
                            status,
                          });
                        }
                      }}
                      style={{ position: 'absolute', inset: 0 }}
                    />
                    <div
                      style={{
                        display: 'flex',
                        alignItems: 'center',
                        gap: spacing.md,
                      }}
                    >
                      <div
                        style={{
                          width: '60px',
                          height: '60px',
                          borderRadius: borderRadius.round,
                          backgroundColor:
                            status === 'completed'
                              ? colors.success
                              : status === 'unlocked'
                              ? colors.secondary
                              : colors.featherLight,
                          display: 'flex',
                          alignItems: 'center',
                          justifyContent: 'center',
                          fontSize: fontSize.xxl,
                          flexShrink: 0,
                        }}
                      >
                        <Icon
                          name={status === 'completed' ? 'check' : status === 'locked' ? 'close' : 'play'}
                          color={status === 'locked' ? colors.earthTone : 'white'}
                          size={28}
                        />
                      </div>
                      <div style={{ flex: 1 }}>
                        <h3
                          style={{
                            fontSize: fontSize.lg,
                            fontWeight: fontWeight.semibold,
                            color: colors.darkBrown,
                            marginBottom: spacing.xs,
                          }}
                        >
                          {lesson.title}
                        </h3>
                        <p
                          style={{
                            fontSize: fontSize.md,
                            color: colors.textLight,
                          }}
                        >
                          {lesson.description}
                        </p>
                        <div
                          style={{
                            marginTop: spacing.sm,
                            fontSize: fontSize.sm,
                            color: colors.earthTone,
                          }}
                        >
                          {lesson.type === 'practice' && 'Practice'}
                          {lesson.type === 'sound' && 'Sound'}
                          {lesson.type === 'song' && 'Song'}
                          {lesson.type === 'echo' && 'Echo'}
                        </div>
                      </div>
                    </div>
                    {status === 'unlocked' && (
                      <div
                        style={{
                          position: 'absolute',
                          top: 0,
                          right: 0,
                          width: '80px',
                          height: '80px',
                          background: `linear-gradient(135deg, transparent 50%, ${colors.warning} 50%)`,
                          display: 'flex',
                          alignItems: 'flex-start',
                          justifyContent: 'flex-end',
                          padding: spacing.xs,
                        }}
                      >
                        <span
                          style={{
                            fontSize: fontSize.xs,
                            fontWeight: fontWeight.bold,
                            color: 'white',
                            transform: 'rotate(45deg) translate(10px, -5px)',
                          }}
                        >
                          NEW
                        </span>
                      </div>
                    )}
                  </Card>
                );
              })}
            </div>
          </div>
        ))}
      </Container>
    </div>
  );
};