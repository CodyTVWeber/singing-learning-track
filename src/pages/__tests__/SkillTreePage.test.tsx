import React from 'react';
import { vi } from 'vitest';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SkillTreePage } from '../SkillTreePage';

vi.mock('../../context/AppContext', () => {
  return {
    useApp: () => ({
      user: {
        id: 'u1',
        name: 'Test User',
        ageGroup: 'kid',
        currentLevel: 1,
        totalPoints: 0,
        streakCount: 1,
        lastStreakDate: null,
      },
      getCompletedLessonIds: () => [],
    }),
  };
});

describe('SkillTreePage', () => {
  it('renders and matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <SkillTreePage />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

