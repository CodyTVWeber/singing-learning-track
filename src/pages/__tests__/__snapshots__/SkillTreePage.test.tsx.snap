// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SkillTreePage > renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="min-height: 100vh; background-color: rgb(250, 247, 242); padding-bottom: 32px;"
  >
    <div
      style="background-color: rgb(139, 105, 82); color: white; padding: 24px; box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06);"
    >
      <div
        class=""
        style="width: 100%; max-width: 1200px; margin: 0px auto; padding: 16px;"
      >
        <div
          style="display: flex; justify-content: space-between; align-items: center; flex-wrap: wrap; gap: 16px;"
        >
          <div>
            <h1
              style="font-size: 24px; font-weight: 700; margin-bottom: 4px; color: white;"
            >
              Hi Test User!
            </h1>
            <p
              style="font-size: 16px; opacity: 0.9;"
            >
              Keep singing every day!
            </p>
          </div>
          <div
            style="display: flex; gap: 24px; align-items: center;"
          >
            <div
              aria-disabled="false"
              aria-pressed="false"
              class=""
              style="display: inline-flex; align-items: center; gap: 4px; height: 32px; padding: 0px 8px; background-color: rgb(255, 183, 77); color: rgb(93, 64, 55); border: 1px solid transparent; border-radius: 50%; font-size: 14px; font-weight: 500; cursor: default; opacity: 1; transition: 150ms ease-in-out; box-shadow: none; user-select: none;"
            >
              <span>
                1 day ðŸ”¥
              </span>
            </div>
            <div
              style="text-align: center;"
            >
              <div
                style="font-size: 12px; opacity: 0.8;"
              >
                Points
              </div>
              <div
                style="font-size: 24px; font-weight: 700;"
              >
                0
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class=""
      style="width: 100%; max-width: 1200px; margin: 0px auto; padding: 16px;"
    >
      <div
        style="margin-top: 32px;"
      >
        <h2
          style="font-size: 32px; font-weight: 700; color: rgb(93, 64, 55); margin-bottom: 16px;"
        >
          Unit 1: Start Singing!
        </h2>
        <p
          style="font-size: 20px; color: rgb(109, 76, 65); margin-bottom: 24px;"
        >
          Learn with Kooka
        </p>
        <div
          style="display: grid; grid-template-columns: repeat(auto-fill, minmax(280px, 1fr)); gap: 24px;"
        >
          <div
            class=""
            style="background-color: rgb(255, 255, 255); border-radius: 16px; padding: 16px; box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06); cursor: pointer; transition: all 300ms ease-in-out; opacity: 1; position: relative; overflow: hidden;"
          >
            <div
              style="position: absolute; inset: 0;"
            />
            <div
              style="display: flex; align-items: center; gap: 16px;"
            >
              <div
                style="width: 60px; height: 60px; border-radius: 50%; background-color: rgb(111, 168, 220); display: flex; align-items: center; justify-content: center; font-size: 32px; flex-shrink: 0;"
              >
                <svg
                  aria-label="play icon"
                  class=""
                  fill="currentColor"
                  style="width: 28px; height: 28px; color: white; cursor: default; transition: 150ms ease-in-out;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8 5v14l11-7z"
                  />
                </svg>
              </div>
              <div
                style="flex: 1 1 0%;"
              >
                <h3
                  style="font-size: 20px; font-weight: 600; color: rgb(93, 64, 55); margin-bottom: 4px;"
                >
                  Echo with Kooka
                </h3>
                <p
                  style="font-size: 16px; color: rgb(109, 76, 65);"
                >
                  Listen and repeat!
                </p>
                <div
                  style="margin-top: 8px; font-size: 14px; color: rgb(161, 136, 127);"
                />
              </div>
            </div>
            <div
              style="position: absolute; top: 0px; right: 0px; width: 80px; height: 80px; background: linear-gradient(135deg, transparent 50%, #FFB74D 50%); display: flex; align-items: flex-start; justify-content: flex-end; padding: 4px;"
            >
              <span
                style="font-size: 12px; font-weight: 700; color: white; transform: rotate(45deg) translate(10px, -5px);"
              >
                NEW
              </span>
            </div>
          </div>
          <div
            class=""
            style="background-color: rgb(255, 255, 255); border-radius: 16px; padding: 16px; box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06); cursor: not-allowed; transition: all 300ms ease-in-out; opacity: 0.6; position: relative; overflow: hidden;"
          >
            <div
              style="position: absolute; inset: 0;"
            />
            <div
              style="display: flex; align-items: center; gap: 16px;"
            >
              <div
                style="width: 60px; height: 60px; border-radius: 50%; background-color: rgb(215, 204, 200); display: flex; align-items: center; justify-content: center; font-size: 32px; flex-shrink: 0;"
              >
                <svg
                  aria-label="close icon"
                  class=""
                  fill="currentColor"
                  style="width: 28px; height: 28px; color: rgb(161, 136, 127); cursor: default; transition: 150ms ease-in-out;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </div>
              <div
                style="flex: 1 1 0%;"
              >
                <h3
                  style="font-size: 20px; font-weight: 600; color: rgb(93, 64, 55); margin-bottom: 4px;"
                >
                  Belly Breathing
                </h3>
                <p
                  style="font-size: 16px; color: rgb(109, 76, 65);"
                >
                  Blow up your belly!
                </p>
                <div
                  style="margin-top: 8px; font-size: 14px; color: rgb(161, 136, 127);"
                >
                  Practice
                </div>
              </div>
            </div>
          </div>
          <div
            class=""
            style="background-color: rgb(255, 255, 255); border-radius: 16px; padding: 16px; box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06); cursor: not-allowed; transition: all 300ms ease-in-out; opacity: 0.6; position: relative; overflow: hidden;"
          >
            <div
              style="position: absolute; inset: 0;"
            />
            <div
              style="display: flex; align-items: center; gap: 16px;"
            >
              <div
                style="width: 60px; height: 60px; border-radius: 50%; background-color: rgb(215, 204, 200); display: flex; align-items: center; justify-content: center; font-size: 32px; flex-shrink: 0;"
              >
                <svg
                  aria-label="close icon"
                  class=""
                  fill="currentColor"
                  style="width: 28px; height: 28px; color: rgb(161, 136, 127); cursor: default; transition: 150ms ease-in-out;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </div>
              <div
                style="flex: 1 1 0%;"
              >
                <h3
                  style="font-size: 20px; font-weight: 600; color: rgb(93, 64, 55); margin-bottom: 4px;"
                >
                  High & Low
                </h3>
                <p
                  style="font-size: 16px; color: rgb(109, 76, 65);"
                >
                  Copy the sounds!
                </p>
                <div
                  style="margin-top: 8px; font-size: 14px; color: rgb(161, 136, 127);"
                >
                  Sound
                </div>
              </div>
            </div>
          </div>
          <div
            class=""
            style="background-color: rgb(255, 255, 255); border-radius: 16px; padding: 16px; box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06); cursor: not-allowed; transition: all 300ms ease-in-out; opacity: 0.6; position: relative; overflow: hidden;"
          >
            <div
              style="position: absolute; inset: 0;"
            />
            <div
              style="display: flex; align-items: center; gap: 16px;"
            >
              <div
                style="width: 60px; height: 60px; border-radius: 50%; background-color: rgb(215, 204, 200); display: flex; align-items: center; justify-content: center; font-size: 32px; flex-shrink: 0;"
              >
                <svg
                  aria-label="close icon"
                  class=""
                  fill="currentColor"
                  style="width: 28px; height: 28px; color: rgb(161, 136, 127); cursor: default; transition: 150ms ease-in-out;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </div>
              <div
                style="flex: 1 1 0%;"
              >
                <h3
                  style="font-size: 20px; font-weight: 600; color: rgb(93, 64, 55); margin-bottom: 4px;"
                >
                  Kooka Laugh Song
                </h3>
                <p
                  style="font-size: 16px; color: rgb(109, 76, 65);"
                >
                  Sing along!
                </p>
                <div
                  style="margin-top: 8px; font-size: 14px; color: rgb(161, 136, 127);"
                >
                  Song
                </div>
              </div>
            </div>
          </div>
          <div
            class=""
            style="background-color: rgb(255, 255, 255); border-radius: 16px; padding: 16px; box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06); cursor: not-allowed; transition: all 300ms ease-in-out; opacity: 0.6; position: relative; overflow: hidden;"
          >
            <div
              style="position: absolute; inset: 0;"
            />
            <div
              style="display: flex; align-items: center; gap: 16px;"
            >
              <div
                style="width: 60px; height: 60px; border-radius: 50%; background-color: rgb(215, 204, 200); display: flex; align-items: center; justify-content: center; font-size: 32px; flex-shrink: 0;"
              >
                <svg
                  aria-label="close icon"
                  class=""
                  fill="currentColor"
                  style="width: 28px; height: 28px; color: rgb(161, 136, 127); cursor: default; transition: 150ms ease-in-out;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </div>
              <div
                style="flex: 1 1 0%;"
              >
                <h3
                  style="font-size: 20px; font-weight: 600; color: rgb(93, 64, 55); margin-bottom: 4px;"
                >
                  Clap the Beat
                </h3>
                <p
                  style="font-size: 16px; color: rgb(109, 76, 65);"
                >
                  Follow the rhythm!
                </p>
                <div
                  style="margin-top: 8px; font-size: 14px; color: rgb(161, 136, 127);"
                >
                  Practice
                </div>
              </div>
            </div>
          </div>
          <div
            class=""
            style="background-color: rgb(255, 255, 255); border-radius: 16px; padding: 16px; box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06); cursor: not-allowed; transition: all 300ms ease-in-out; opacity: 0.6; position: relative; overflow: hidden;"
          >
            <div
              style="position: absolute; inset: 0;"
            />
            <div
              style="display: flex; align-items: center; gap: 16px;"
            >
              <div
                style="width: 60px; height: 60px; border-radius: 50%; background-color: rgb(215, 204, 200); display: flex; align-items: center; justify-content: center; font-size: 32px; flex-shrink: 0;"
              >
                <svg
                  aria-label="close icon"
                  class=""
                  fill="currentColor"
                  style="width: 28px; height: 28px; color: rgb(161, 136, 127); cursor: default; transition: 150ms ease-in-out;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </div>
              <div
                style="flex: 1 1 0%;"
              >
                <h3
                  style="font-size: 20px; font-weight: 600; color: rgb(93, 64, 55); margin-bottom: 4px;"
                >
                  Animal Voices
                </h3>
                <p
                  style="font-size: 16px; color: rgb(109, 76, 65);"
                >
                  Be the animals!
                </p>
                <div
                  style="margin-top: 8px; font-size: 14px; color: rgb(161, 136, 127);"
                >
                  Sound
                </div>
              </div>
            </div>
          </div>
          <div
            class=""
            style="background-color: rgb(255, 255, 255); border-radius: 16px; padding: 16px; box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06); cursor: not-allowed; transition: all 300ms ease-in-out; opacity: 0.6; position: relative; overflow: hidden;"
          >
            <div
              style="position: absolute; inset: 0;"
            />
            <div
              style="display: flex; align-items: center; gap: 16px;"
            >
              <div
                style="width: 60px; height: 60px; border-radius: 50%; background-color: rgb(215, 204, 200); display: flex; align-items: center; justify-content: center; font-size: 32px; flex-shrink: 0;"
              >
                <svg
                  aria-label="close icon"
                  class=""
                  fill="currentColor"
                  style="width: 28px; height: 28px; color: rgb(161, 136, 127); cursor: default; transition: 150ms ease-in-out;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </div>
              <div
                style="flex: 1 1 0%;"
              >
                <h3
                  style="font-size: 20px; font-weight: 600; color: rgb(93, 64, 55); margin-bottom: 4px;"
                >
                  Echo Echo
                </h3>
                <p
                  style="font-size: 16px; color: rgb(109, 76, 65);"
                >
                  Copy me!
                </p>
                <div
                  style="margin-top: 8px; font-size: 14px; color: rgb(161, 136, 127);"
                >
                  Practice
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
